File: Java/Silver/1874.java
import java.io.*;
import java.util.*;

public class Main {

    private static final BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    private static final StringBuilder sb = new StringBuilder();

    public static void main(String[] args) throws IOException {

        int n = Integer.parseInt(br.readLine());

        checkStackSequence(n);
        br.close();
    }

    private static void checkStackSequence(int n) throws IOException {

        Queue<Integer> queue = new ArrayDeque<>();
        for (int i = 1; i <= n; i++){ queue.offer(i); }

        ArrayDeque<Integer> stack = new ArrayDeque<>();

        while (n-- > 0) {
            int num = Integer.parseInt(br.readLine());

            while (stack.isEmpty() || stack.peekFirst() < num) {
                stack.offerFirst(queue.poll());
                sb.append("+\n");
            }

            if (stack.peekFirst() == num) {
                stack.pollFirst();
                sb.append("-\n");
                continue;
            }

            if (stack.peekFirst() > num) {
                System.out.println("NO");
                return;
            }

        }

        System.out.println(sb);
    }
}

File: Archived2/Kotlin/Silver/1874.kt
import java.io.BufferedReader
import java.io.InputStreamReader


fun main() {

    val br = BufferedReader(InputStreamReader(System.`in`))
    checkStackSequence(br)
    br.close()

}

fun checkStackSequence(br: BufferedReader) {

    val n = br.readLine().toInt()
    val queue = ArrayDeque<Int>()
    repeat (n) { index -> queue.addLast(index + 1)}

    val stack = ArrayDeque<Int>()
    val answer = buildString {

        repeat (n) {

            val num = br.readLine().toInt()

            while (stack.isEmpty() || stack.component1() < num) {
                append("+\n")
                stack.addFirst(queue.removeFirst())
            }

            if (stack.component1() > num) {
                println("NO")
                return
            }

            if (stack.component1() == num) {
                append("-\n")
                stack.removeFirst()
            }

        }
    }

    println(answer)

}